### PR TITLE
Type DatasetFactory

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
@fjetter Does this need a changelog entry?

File: kartothek/core/factory.py [hunk 1]
```diff
@@ -39,12 +45,12 @@ class DatasetFactory(DatasetMetadataBase):
...
     def __init__(
         self,
-        dataset_uuid,
-        store_factory,
-        load_schema=True,
-        load_all_indices=False,
-        load_dataset_metadata=True,
-    ):
+        dataset_uuid: str,
+        store_factory: Callable[[], "KeyValueStore"],
```
Comment (lr4d) [inline kartothek/core/factory.py]: `store_factory()` is called a bit down, wouldn't this fail for `KeyValueStore`?

Comment (xhochy) [reply kartothek/core/factory.py]: We call `_check_callable` which should fail for just `KeyValueStore`.

Comment (lr4d) [reply kartothek/core/factory.py]: my bad, didn't read the types properly

**Comment by xhochy** [comment]
pre-commit mystery again: locally they pass.

**Comment by xhochy** [comment]
Mystery solved: The hooks run in CI with Python 3.6 and locally with Python 3.7. I have used Python 3.7+ features and thus some typing things are not correctly resolved in the CI.

**Comment by fjetter** [comment]
> Does this need a changelog entry?

No

**Pull request merged**